ports/nrf52840/core/hardware/bma400/bma400.cpp

```diff
@@ -0,0 +1,397 @@
+
+    /*! Read write length varies based on user requirement */
+    #define READ_WRITE_LENGTH  UINT8_C(46) // todo: get the required read-write length
+    m_bma400_dev.read_write_len = READ_WRITE_LENGTH;
+
+    int8_t err_code;
+
+    err_code = bma400_init(&m_bma400_dev);
+
+    if (err_code != BMA400_OK)
+    {
+        DEBUG_ERROR("BMA400 initialization failure (%d)",err_code);
+        throw ErrorCode::I2C_COMMS_ERROR;
+    }
+    // rslt = bma400_soft_reset(&m_bma400_dev);
+    // bma400_check_rslt("bma400_soft_reset", rslt);
+
+    // rslt = bma400_init(&m_bma400_dev);
+    // bma400_check_rslt("bma400_init", rslt);
+
+    /* Get the accelerometer configurations which are set in the sensor */
+    rslt = bma400_get_sensor_conf(&m_bma400_sensor_conf, 1, &m_bma400_dev);
+    bma400_check_rslt("bma400_get_sensor_conf", rslt);
+
+    /* Select the type of configuration to be modified */
+    m_bma400_sensor_conf.type = BMA400_ACCEL;
+    /* Modify the desired configurations as per macros
+     * available in bma400_defs.h file */
+    m_bma400_sensor_conf.param.accel.odr = BMA400_ODR_100HZ;
+    m_bma400_sensor_conf.param.accel.range = BMA400_RANGE_16G;
+    m_bma400_sensor_conf.param.accel.data_src = BMA400_DATA_SRC_ACCEL_FILT_2;
+
+    /* Set the desired configurations to the sensor */
+    rslt = bma400_set_sensor_conf(&m_bma400_sensor_conf, 1, &m_bma400_dev);
+    bma400_check_rslt("bma400_set_sensor_conf", rslt);
+
+    // note: Sleep mode: Registers readable and writable, no sensortime
+    rslt = bma400_set_power_mode(BMA400_MODE_SLEEP, &m_bma400_dev); 
+    bma400_check_rslt("bma400_set_power_mode", rslt);
+
+
+    m_bma400_device_conf.type = BMA400_INT_PIN_CONF;
+    m_bma400_device_conf.param.int_conf.int_chan = BMA400_INT_CHANNEL_1;
+    m_bma400_device_conf.param.int_conf.pin_conf = BMA400_INT_PUSH_PULL_ACTIVE_0;
+    rslt = bma400_set_device_conf(&m_bma400_device_conf, BMA400_INT_PIN_CONF, &m_bma400_dev);
+    bma400_check_rslt("bma400_set_device_conf", rslt);
+}
+//                       
+int8_t BMA400LL::i2c_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t length, void *intf_ptr)
+{
+    nrfx_err_t err_code;
+    BMA400LL& device = BMA400LLManager::lookup_device(0); //todo  Define intf_ptr to include chip_id
+
+    if (!length)
+        return BMA400_OK;
+    
+    if (length > BMA400_MAX_BUFFER_LEN + sizeof(reg_addr))
+        return BMA400_E_OUT_OF_RANGE;
+    
+    uint8_t buffer[BMA400_MAX_BUFFER_LEN];
+    buffer[0] = reg_addr;
+    memcpy(&buffer[1], reg_data, length);
+    
+    nrfx_twim_xfer_desc_t xfer = NRFX_TWIM_XFER_DESC_TX(device.m_addr, (uint8_t*)buffer, length + sizeof(reg_addr));
+
+    err_code = nrfx_twim_xfer(&BSP::I2C_Inits[device.m_bus].twim, &xfer, 0);
+    //todo: what does this need?
+//    err_code = nrfx_twim_tx(&BSP::I2C_Inits[device.m_bus].twim, device.m_addr, buffer, length + sizeof(reg_addr), false);
+    if (err_code != NRFX_SUCCESS)
+        return BMA400_E_COM_FAIL;
```

Comment:
I recommend using the NrfI2C wrapper class.  See nrf_i2c.hpp.  Not sure why the old driver isn't using it but most i2c stuff is now moved over to the wrapper; example: https://github.com/arribada/CLS-Argos-Linkit-CORE/blob/master/ports/nrf52840/core/hardware/ltr_303/ltr_303.cpp.

It handles the exception throwing directly so you don't need your switch table for error checking.  Should save some code.

Having said that, I do note that the existing code uses `twim_tx` and not `twim_xfer` although it should make no difference judging by your descriptor.